e2e_playwright/hostframe_app_test.py

```diff
@@ -0,0 +1,187 @@
+
+    # Open the Main Menu
+    frame_locator.get_by_test_id("stMainMenu").locator("button").click()
+    # Open the Settings Modal
+    frame_locator.get_by_test_id("stMainMenuList").get_by_text("Settings").click()
+
+    expect(frame_locator.get_by_role("dialog")).to_be_attached()
+    # Close the Modal
+    toolbar_buttons.get_by_text("Close Modal").click()
+    expect(frame_locator.get_by_role("dialog")).not_to_be_attached()
+
+
+def test_handles_host_menu_item_message(iframed_app: IframedPage):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    toolbar_buttons.get_by_text("Add Menu Item").click()
+    # Open the Main Menu
+    frame_locator.get_by_test_id("stMainMenu").locator("button").click()
+    menu_list = frame_locator.get_by_test_id("stMainMenuList")
+    expect(menu_list.get_by_text("Adopt a Corgi")).to_be_attached()
+
+
+def test_handles_host_toolbar_item_message(
+    iframed_app: IframedPage, assert_snapshot: ImageCompareFunction
+):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    toolbar_buttons.get_by_text("Add Toolbar Item").click()
+
+    toolbar_action_button = frame_locator.get_by_test_id("stToolbarActionButton")
+    expect(toolbar_action_button).to_have_count(2)
+    expect(toolbar_action_button.nth(0)).to_have_text("Favorite")
+    expect(toolbar_action_button.nth(1)).to_have_text("Share")
+
+    assert_snapshot(
+        frame_locator.get_by_test_id("stApp"), name="hostframe_app-toolbar_items"
+    )
+
+
+def test_handles_hide_sidebar_nav_message(iframed_app: IframedPage):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    expect(frame_locator.get_by_test_id("stSidebarNav")).to_be_attached()
+    toolbar_buttons.get_by_text("Hide Sidebar Nav").click()
+    expect(frame_locator.get_by_test_id("stSidebarNav")).not_to_be_attached()
+
+
+def test_handles_sidebar_downshift_message(iframed_app: IframedPage):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    # Trigger sidebar downshift
+    toolbar_buttons.get_by_text("Sidebar Chevron Downshift").click()
+    # Hover over sidebar to reveal the button
+    frame_locator.get_by_test_id("stSidebarContent").hover()
+    # Close the sidebar
+    frame_locator.get_by_test_id("stSidebar").locator("button").click()
+    # Check chevron positioning
+    expect(frame_locator.get_by_test_id("stSidebarCollapsedControl")).to_have_css(
+        "top", "50px"
+    )
+
+
+def test_handles_host_terminate_and_restart_websocket_connection_messages(
+    iframed_app: IframedPage,
+):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+
+    # Kill the websocket connection and verify that the app moves into an
+    # error state.
+    toolbar_buttons.get_by_text("Terminate Websocket").click()
+    expect(frame_locator.get_by_test_id("stStatusWidget")).to_have_text("Error")
```

Comment:
nit: is the error status also reflected on one of the `stApp` attributes? maybe its better to rely on those attributes instead of the `stStatusWidget`

Reply:
Maybe it is represented in  `data-test-connection-state==DISCONNECTED_FOREVER`  

```
expect(frame_locator.get_by_test_id("stApp")).to_have_attribute("data-test-connection-state", "DISCONNECTED_FOREVER")
```

Reply:
yep, good point!